jahsieAult/homework2Answers.md

```diff
@@ -0,0 +1,308 @@
+
+var song;
+var newGift = [];
+
+
+for (i = 0; i < whatDay.length; i++) { //creates loop
+  if (i < 1) {
+    song = "On the " + whatDay[i] + startVerse + "A " + gift[i] + "\n"
+  }
+  else {
+    newGift.unshift(gift[i]);
+    song += "On the " + whatDay[i] + startVerse + newGift.join("") + "And a " + gift[0] + "\n"; 
+  }
+}
+console.log(song);
+}
+writeSong()
+```
+---
+
+**4)** _(Difficult, 20%)_
+
+Suppose the '&' key on your keyboard is missing, and you want to be able to continue programming without it.  You decide to simulate the && operator with a function!
+
+**a)**
+Write a function `and2(a,b)` which tries to simulate the && operator: it should always return the same result as `(a && b)` for any values of _a_ and _b_.  (For example, `and2((0>1),true)` should return _false_.)  But you can't use && itself within your function!
+
+Answer:
+(I couldn't figure out how to indent this properly.)
+
+```
+function and2(a,b) {
+	if (a || b) {
+		if (a) {
+			if (b) {
+				return true;
+			}
+			else {
+				return false;
+      		}
+     	}
+		else {
+			return false;
+			}	
+		}
+  	else return false;
+	}
+
+```
+
+**b)** Write another function `and3(a,b,c)` which tries to simulate a double-&& operator: it should always return the same result as `(a && b && c)` for any values of a,b,c.  (For example, `and3((1>0),(0>1),true)` should return _false_.)  As before, you're not allowed to use && itself. Remember that you can call functions from functions!
+
+Answer: 
+
+```
+function and3(a,b,c) {
+		if (and2(a, b) === false) {
+			return false;
+		}
+		else if (c) {
+			return true;
+		}
+      else {
+      	return false;
+      }
+	}
+
+```
+
+**c)**
+Now generalize your function to handle any number of values.  You will learn better ways eventually, but for now use an array to store all the values.
+Your new function `andN(values)` should accept one parameter _values_, which is an array holding all the values to && together.
+If the argument you provide when calling the function holds values [a,b,c...z], as in
+`andN(26,[a,b,c...z])`, the function should return the same result as `(a && b && c && ... z)`.
+Make sure to handle two special cases: length 0 (then return _true_) and length 1 (then return that single value).
+Again, don't use &&.
+
+Answer:
+
+(I couldn't get length 0 to return true, it kept coming up as undefined.)
+
+```
+function andN(values) {
+	for (i = 0; i < values.length; i++) {
+		if (values.length === 0) {
+			return true;
+			}
+			else if (values.length === 1) {
+				return values[0];
+			}
+				else if (values[i] === false) {
+					return false;
+				}
+          else if (values[values.length -1 ] === true) {
+                  return true;}
+          }
+          
+}
```

Comment:
4c) I had to tinker a little bit to figure out why you weren't returning true for values.length === 0. The problem is in the for loop definition: the loop will only run while i < values.length, so if there are no values, the loop doesn't run at all, meaning that it never gets to the if statement.
The other tricky bit in here is the behavior of `return`: Any time that runs, it stops the loop. This is problematic for your solution in your last if statement; it checks the last element in the array, and if it's true, it finishes the loop without having gone through any of the other elements other than the first one.